### ISSUE TITLE
OCommandSQLParsingException on 'SELECT' with 'LIMIT' and 'LOCK'

### ISSUE THREAD

**Issue by matanshukry** [orientechnologies/orientdb#4744]
Consider the following queries:
1. ```SELECT from USER LIMIT 1 LOCK RECORD```
2. ```SELECT from USER LOCK RECORD LIMIT 1``` - fail
3. ```UPDATE USER SET name="Jay" LIMIT 1 LOCK RECORD``` - fail
4. ```UPDATE USER SET name="Jay" LOCK RECORD LIMIT 1```
5. ```UPDATE USER SET name="Jay" WHERE id=0 LOCK RECORD LIMIT 1``` - fail
6. ```UPDATE USER SET name="Jay" WHERE id=0 LIMIT 1 LOCK RECORD``` - fail
7. ```UPDATE USER SET name="Jay" LOCK RECORD LIMIT 1 WHERE id=0``` - fail
8. ```UPDATE USER SET name="Jay" LIMIT 1 LOCK RECORD WHERE id=0``` - fail
9. ```UPDATE USER SET name="Jay" LIMIT 1 WHERE id=0 LOCK RECORD``` - fail
10. ```UPDATE USER SET name="Jay" LOCK RECORD WHERE id=0 LIMIT 1``` - fail
```

Queries 2,3, and 5-10 will all fail with similar exceptions to this:
```
Error: com.orientechnologies.orient.core.sql.OCommandSQLParsingException: Error on parsing command at position #0: Encountered " <LOCK> "LOCK "" at line 1, column 47.
Was expecting one of:
    <EOF> 
   <TIMEOUT> ...
    ";" ...
```

When looking at the first SELECT queries it seems like a small issue, since one can simply switch the clause order; However, when looking at the other queries, one can see a combination of ```LOCK```, ```LIMIT``` and ```WHERE``` is not possible, since all will result in an exception.

I tried tracking the problem down, but it got me to OrientSql.java - which is the not-so-nice parser code. Perhaps the one who wrote it might have better luck messing around with that file.

p.s.
As far as I can tell, there is nothing in the documentation that states one can use the clauses in any order. However, in practice, I found out the order usually doesn't matter.
Perhaps it should be added to documentation, assuming the order really doesn't matter?

**Comment by nagarajasr** [comment]
@matanshukry OrientSql.java is generated using javacc. the grammar is specified in core/src/main/grammar/OrientSQL.jjt which is not too hard to follow. the parser validates the syntax before submitting the command to backend executor. in some cases the executor enforces the order of clauses even if the parser lets it through. in any case, the executor is being rewritten in 2.2 perhaps a good time to set this in order, IMO.

**Comment by matanshukry** [comment]
Can you elaborate on future plans?
Will the parser be rewritten as well?
Will it still use JJTree?
If not, what then?

**Comment by luigidellaquila** [comment]
Hi Matan,

the parser has already been rewritten (with JJTree), but after parsing, the
old executor is still running.
Next step is to also rewrite the execution plan and execution itself.

Luigi


2015-08-07 21:36 GMT+02:00 Matan Shukry <notifications@github.com>:

**Comment by matanshukry** [comment]
I am closing this issue then. I'll rewrite in the future in case of any problems, although I don't expect any.
